src/core/mapping/base_mapper.cc

```diff
@@ -767,36 +757,75 @@ void BaseMapper::report_failed_mapping(const Legion::Mappable& mappable,
-               get_mapper_name(),
+    req_ss << "region requirement(s) ";
+  bool past_first = false;
+  for (uint32_t index : mapping.requirement_indices()) {
+    if (past_first)
+      req_ss << ",";
+    else
+      past_first = true;
+    req_ss << index;
+  }
+
+  logger.error("Out of memory: Legate failed to allocate %zd bytes on memory " IDFMT
+               " (of kind %s) for %s of %s%s[%s] (UID %lld)",
                footprint,
                target_memory.id,
                Legion::Mapping::Utilities::to_string(target_memory.kind()),
-               memory_kinds[target_memory.kind()],
                req_ss.str().c_str(),
                log_mappable(mappable, true /*prefix_only*/).c_str(),
                opname.c_str(),
                provenance.c_str(),
                mappable.get_unique_id());
+
+  auto log_field_info = [&](Legion::FieldSpace fs, const auto& fields) {
+    size_t size;
+    const void* alloc_info;
+    for (Legion::FieldID fid : fields)
+      if (runtime->retrieve_semantic_information(
+            ctx, fs, fid, LEGATE_CORE_ALLOC_INFO_TAG, alloc_info, size, true /* can_fail */))
+        logger.error() << "corresponding to a Store allocated at "
+                       << static_cast<const char*>(alloc_info);
+  };
+  for (const Legion::RegionRequirement* req : mapping.requirements())
+    log_field_info(req->region.get_field_space(), req->instance_fields);
+
+  std::vector<Legion::Mapping::PhysicalInstance> existing;
+  runtime->find_physical_instances(ctx,
+                                   target_memory,
+                                   Legion::LayoutConstraintSet(),
+                                   std::vector<Legion::LogicalRegion>(),
+                                   existing);
+  size_t total_size = 0;
+  for (Legion::Mapping::PhysicalInstance inst : existing) total_size += inst.get_instance_size();
+  logger.error() << "There is not enough space because Legate is reserving " << total_size
+                 << " of the available " << target_memory.capacity() << " bytes (minus the eager"
+                 << " pool allocation) for the following Instances:";
+  for (Legion::Mapping::PhysicalInstance inst : existing) {
+    logger.error() << Legion::Mapping::Utilities::to_string(runtime, ctx, inst) << " of size "
+                   << inst.get_instance_size();
+    std::set<Legion::FieldID> fields;
+    inst.get_fields(fields);
+    log_field_info(inst.get_field_space(), fields);
+  }
+
+  logger.error() << "Here are some things to try:";
+  logger.error() << "* Make sure your code is not impeding the garbage collection of Legate-backed "
+                 << "objects, e.g. by storing references in caches, or creating reference cycles.";
+  logger.error() << "* Ask Legate to reserve more space on the above memory, using the appropriate "
+                 << "--*mem legate flag.";
+  logger.error() << "* Assign less memory to the eager pool, by reducing --eager-alloc-percentage.";
```

Comment:
super nit: "eager pool by reducing" with no comma. There's no emphasized pause. Kind of looks like you are listing elements.